File: prompt_hr/py/employee_onboarding.py
import frappe
from prompt_hr.py.utils import send_notification_email

# ! prompt_hr.py.employee_onboarding.get_onboarding_details
# ? FETCH TEMPLATE ACTIVITIES FOR CHILD TABLE
@frappe.whitelist()
def get_onboarding_details(parent, parenttype):
    return frappe.get_all(
        "Employee Boarding Activity",
        fields=["*"],
        filters={"parent": parent, "parenttype": parenttype},
        order_by="idx"
    )

# ? AFTER INSERT EVENT
def after_insert(doc, method):
    # ? SET REQUIRED DOCUMENTS IN THE NEW JOINEE CHECKLIST
    # set_required_documents_in_new_joinee_checklist(doc)
    pass

# ? FUNCTION TO SET REQUIRED DOCUMENTS IN THE NEW JOINEE CHECKLIST
def set_required_documents_in_new_joinee_checklist(doc):
    if doc.job_applicant:
        # ? GET JOINING DOCUMENT CHECKLIST
        joining_document_checklist = frappe.get_value(
            "Joining Document Checklist",
            {"company": doc.company},
            "name"
        )

        # ? GET JOB APPLICANT'S REQUIRED DOCUMENTS
        documents = frappe.get_all(
            "Joining Document",
            filters={
                "parent": joining_document_checklist,
                "document_collection_stage": "Employee Onboarding"
            },
            fields=["required_document", "document_collection_stage"]
        )

        new_joinee_checklist = frappe.get_doc("New Joinee Checklist", {"job_applicant": doc.job_applicant})
        if not new_joinee_checklist:
            return

        for doc_item in documents:
            new_joinee_checklist.append("required_documents", {
                "required_document": doc_item.required_document,
                "collection_stage": doc_item.document_collection_stage
            })

        new_joinee_checklist.save(ignore_permissions=True)
        frappe.db.commit()

        


# ? MAIN ON_UPDATE EVENT FUNCTION
@frappe.whitelist()
def validate(doc, method):

    # auto_fill_first_activity(doc)
    fill_missing_checklist_records(doc)
    rows_to_notify = get_pending_activity_rows(doc)
    frappe.db.commit()

    notify_users_for_pending_actions(rows_to_notify, company=doc.company)

    return "Emails enqueued successfully"


# ? FILL FIRST ROW USER AND CHECKLIST RECORD IF EMPTY
def auto_fill_first_activity(doc):
    if not doc.activities:
        return 

    first = doc.activities[0]

    if not first.user and not first.custom_checklist_record and doc.job_applicant:
        email = get_applicant_email(doc.job_applicant)
        checklist_record = get_checklist_record("New Joinee Checklist", doc.job_applicant)

        if email:
            first.user = email
            first.custom_checklist_record = checklist_record

            # ? ONLY SEND IF IS_RAISED IS ALREADY SET
            if first.custom_is_sent == 0:
                send_pending_action_email(first, notification_name="Onboarding Activity Reminder")
                first.custom_is_sent = 1


# ? GET EMAIL FROM JOB APPLICANT
def get_applicant_email(job_applicant):
    return frappe.get_value("Job Applicant", job_applicant, "email_id")


# ? FETCH CHECKLIST RECORD BY DOCTYPE NAME AND JOB APPLICANT
def get_checklist_record(doctype_name, job_applicant):
    try:
        checklist_record_name = frappe.get_value(doctype_name, {"job_applicant": job_applicant}, "name")
        if not checklist_record_name:
            checklist_record = frappe.new_doc(doctype_name)
            checklist_record.job_applicant = job_applicant
            checklist_record.insert(ignore_permissions=True)
            checklist_record_name = checklist_record.name
        else:
            print(f"[DEBUG] Found existing checklist record: {checklist_record_name}")
        return checklist_record_name
    except Exception as e:
        frappe.log_error(f"Error fetching checklist from {doctype_name}: {e}", "Checklist Fetch Error")
        return None

# ? FILL MISSING CHECKLIST RECORDS IN ACTIVITIES
def fill_missing_checklist_records(doc):
    for row in doc.activities:
        if not row.custom_checklist_record and row.custom_checklist_name and doc.job_applicant:
            checklist_record = get_checklist_record(row.custom_checklist_name, doc.job_applicant)
            if checklist_record:
                row.custom_checklist_record = checklist_record


# ? GET FILTERED ROWS WHERE EMAIL SHOULD BE SENT
def get_pending_activity_rows(doc):
    filtered = [
        row for row in doc.activities
        if (row.user or row.role) and row.custom_is_raised == 1 and row.custom_is_sent == 0
    ]

    return filtered


# ? SEND EMAIL TO USERS FOR PENDING CHECKLIST ACTIONS
def notify_users_for_pending_actions(rows, company):
    if not rows:
        return
    for row in rows:
        send_pending_action_email(row, notification_name="Reporting Manger Checklist", company=company)
        row.custom_is_sent = 1  

# ? FUNCTION TO COMPOSE + SEND EMAIL FOR A SINGLE ROW
def send_pending_action_email(row, notification_name, company):
    # ? FETCH DOCTYPE AND DOCNAME
    doc_type = row.custom_checklist_name
    doc_name = row.custom_checklist_record
    recipient = row.user

    # ? VALIDATE IF BOTH USER AND ROLE ARE SET
    if row.user and row.role:
        frappe.throw("Kindly Either Select User or Role, Not Both", title="Invalid Selection")

    # ? IF ROLE IS SELECTED
    if row.role:
        # ? GET USERS WITH THE SPECIFIED ROLE
        users = frappe.get_all(
            "Has Role",
            filters={"role": row.role},
            fields=["parent as user"],
            pluck="user"
        )

        if not users:
            frappe.throw(f"No users found with role {row.role}", title="No Users Found")

        # ? FILTER ACTIVE EMPLOYEES FROM THE USERS
        emp_wise_recipient = frappe.get_all(
            "Employee",
            filters={
                "user_id": ["in", users],
                "company": company,
                "status": "Active"
            },
            fields=["user_id"]
        )

        if not emp_wise_recipient:
            frappe.throw(f"No active employees found with role {row.role}", title="No Active Employees Found")

        # ? EXTRACT USER IDS
        user_ids = [emp.user_id for emp in emp_wise_recipient if emp.user_id]
        if not user_ids:
            frappe.throw(f"No valid user IDs found for employees with role {row.role}", title="No Valid Users")

        # ? GET EMAIL ADDRESSES OF VALID USERS
        recipients = frappe.get_all(
            "User",
            filters={
                "name": ["in", user_ids],
                "enabled": 1
            },
            fields=["email"],
            pluck="email"
        )

        # ? FILTER OUT EMPTY EMAILS
        recipients = [email for email in recipients if email]

        if not recipients:
            frappe.throw(f"No email addresses found for users with role {row.role}", title="No Email Addresses")

    # ? IF SINGLE USER IS SELECTED
    else:
        if not recipient:
            frappe.throw("No user or role specified", title="Missing Recipient")

        # ? GET EMAIL FOR SINGLE USER
        user_email = frappe.db.get_value("User", recipient, "email")
        if not user_email:
            frappe.throw(f"No email found for user {recipient}", title="No Email Found")

        recipients = [user_email]

    # ? SEND NOTIFICATION EMAIL
    send_notification_email(
        recipients=recipients,
        notification_name=notification_name,
        doctype=doc_type,
        docname=doc_name,
        button_label="View Details",
    )

    frappe.msgprint("Notification email sent successfully.")

File: prompt_hr/public/js/employee_onboarding.js
frappe.ui.form.on("Employee Onboarding", {
    
    // ? RUN ON FORM REFRESH
    refresh: async function (frm) {
        // update_first_activity_if_needed(frm);
        let applicants = await frappe.db.get_list("Job Applicant", {
            fields: ["name"],
            or_filters: [
                ["status", "=", "Accepted"],
                ["custom_is_offer_required_while_onboarding", "=", 0]
            ],
        });
        frm.set_query("job_applicant", function () {
            return {
                filters: [
                    ["name", "in",applicants.map(applicant => applicant.name)],
                ]
            };
        });
        if (frm.doc.job_applicant) {
            frappe.db.get_doc("Job Applicant", frm.doc.job_applicant).then(applicant_data => {
                if (!applicant_data.custom_is_offer_required_while_onboarding) {
                    frm.toggle_reqd("job_offer", 0);
                } else {
                    frm.toggle_reqd("job_offer", 1);
                }
            });
        }

        if (!frm.is_new() && frm.doc.employee_onboarding_template) {

            // ? INVITE FOR DOCUMENT COLLECTION BUTTON
            createInviteButton(frm);
        }
    },
    // ? WHEN ONBOARDING TEMPLATE IS SELECTED
    employee_onboarding_template: function (frm) {
        handle_template_selection(frm);
    },
    job_applicant: function (frm) {
        if (frm.doc.job_applicant) {
            frappe.db.get_doc("Job Applicant", frm.doc.job_applicant).then(applicant_data => {
                if (!applicant_data.custom_is_offer_required_while_onboarding) {
                    frm.toggle_reqd("job_offer", 0);
                } else {
                    frm.toggle_reqd("job_offer", 1);
                }
            });
        }
    }
});

// ? HANDLE TEMPLATE SELECTION: CLEAR + FETCH + POPULATE
function handle_template_selection(frm) {
    clear_activities(frm);

    if (frm.doc.employee_onboarding_template) {
        fetch_template_activities(frm.doc.employee_onboarding_template, function (activities) {
            populate_activities(frm, activities);
        });
    }
}

// ? CLEAR EXISTING ACTIVITIES CHILD TABLE
function clear_activities(frm) {
    frm.set_value("activities", "");
}

// ? FETCH ACTIVITIES FROM SERVER BASED ON TEMPLATE
function fetch_template_activities(template_name, callback) {
    frappe.call({
        method: "prompt_hr.py.employee_onboarding.get_onboarding_details",
        args: {
            parent: template_name,
            parenttype: "Employee Onboarding Template",
        },

        // ? WHEN DATA IS RETURNED
        callback: function (r) {
            if (r.message) {
                callback(r.message);
            } else {
                callback([]);
            }
        }
    });
}

// ? POPULATE ACTIVITIES CHILD TABLE AND REFRESH FIELD
function populate_activities(frm, activities) {
    activities.forEach((d) => {
        frm.add_child("activities", d);
    });
    refresh_field("activities");
}

// ? SET FIRST ACTIVITY custom_is_raised TO 1 IF IT'S 0
function update_first_activity_if_needed(frm) {
    const first_activity = frm.doc.activities && frm.doc.activities[0];

    if (first_activity && first_activity.custom_is_raised == 0) {

        // ? UPDATE VALUE
        first_activity.custom_is_raised = 1;

        // ? REFRESH CHILD TABLE
        refresh_field("activities");

        // ? SAVE FORM
        frm.save();
    }
}

// ? CREATE INVITE FOR DOCUMENT COLLECTION BUTTON
function createInviteButton(frm) {
    frm.add_custom_button(__('Invite for Document Collection'), function () {

        // ? CREATE A DIALOG TO COLLECT EXTRA DETAILS
        let dialog = new frappe.ui.Dialog({
            title: 'Invite for Document Collection',
            fields: [
                {
                    label: 'Joining Document Checklist',
                    fieldname: 'joining_document_checklist',
                    fieldtype: 'Link',
                    options: 'Joining Document Checklist',
                    reqd: 1,
                    onchange: function () {

                        // ?  FETCH DOCUMENTS FROM THE SELECTED CHECKLIST USING A CUSTOM METHOD
                        if (dialog.get_value('joining_document_checklist')) {
                            dialog.set_df_property('documents', 'data', []);

                            frappe.call({
                                method: "prompt_hr.py.utils.get_checklist_documents",
                                args: {
                                    checklist: dialog.get_value('joining_document_checklist')
                                },
                                callback: function (r) {

                                    // ?  HIDE LOADING INDICATOR
                                    frappe.hide_progress();

                                    // ?  DEBUG RESPONSE
                                    console.log("API Response:", r);

                                    // ?  CHECK FOR ERROR IN RESPONSE
                                    if (r.exc) {
                                        frappe.msgprint({
                                            title: __("Error"),
                                            message: __("Error fetching documents: ") + r.exc,
                                            indicator: "red"
                                        });
                                        return;
                                    }
                                    // ?  VALIDATE DOCUMENTS DATA
                                    if (r.message && r.message.documents) {
                                        let documents = r.message.documents;
                                        console.log("Documents received:", documents);

                                        if (!documents.length) {
                                            frappe.msgprint({
                                                title: __("Info"),
                                                message: __("No documents found in the selected checklist."),
                                                indicator: "blue"
                                            });
                                            return;
                                        }
                                        // ?  UPDATE THE DATA FOR THE DOCUMENTS FIELD
                                        dialog.fields[3].data = documents;

                                        dialog.fields_dict.documents.grid.refresh();
                                    } else {
                                        console.error("Invalid response format:", r.message);
                                        frappe.msgprint({
                                            title: __("Error"),
                                            message: __("Invalid response format from server."),
                                            indicator: "red"
                                        });
                                    }
                                }
                            });
                        }
                    }
                },
                {
                    label: 'Document Collection Stage',
                    fieldname: 'document_collection_stage',
                    fieldtype: 'Link',
                    options: 'Document Collection Stage',
                    onchange: function () {
                        // ? UPDATE ALL ROWS WITH THE SELECTED STAGE
                        let stage = dialog.get_value('document_collection_stage');
                        if (stage) {

                            let documents = dialog.fields[3].data
                            let newDocuments = documents.filter(doc => {
                                return doc.document_collection_stage == stage;
                            });

                            // ?  UPDATE THE DATA FOR THE DOCUMENTS FIELD
                            dialog.fields[3].data = newDocuments;
                            dialog.fields_dict.documents.grid.refresh();
                        }
                    }
                },
                {
                    label: 'Documents',
                    fieldname: 'documents',
                    fieldtype: 'Table',
                    cannot_add_rows: true,
                    fields: [
                        {
                            label: 'Required Document',
                            fieldname: 'required_document',
                            fieldtype: 'Link',
                            options: 'Required Document Applicant',
                            in_list_view: 1,
                            read_only: 1
                        },
                        {
                            label: 'Document Collection Stage',
                            fieldname: 'document_collection_stage',
                            fieldtype: 'Link',
                            options: 'Document Collection Stage',
                            in_list_view: 1,
                            read_only: 1
                        }
                    ],
                    data: []
                }
            ],
            primary_action_label: 'Send Invite',
            primary_action(values) {

                // ?  VALIDATE THAT WE HAVE DOCUMENTS
                if (!dialog.fields_dict.documents.grid.grid_rows || !dialog.fields_dict.documents.grid.grid_rows.length) {
                    frappe.msgprint({
                        title: __("Warning"),
                        message: __("No documents selected for collection. Please select a valid checklist."),
                        indicator: "yellow"
                    });
                    return;
                }
                dialog.hide();

                // ?  GET SELECTED DOCUMENTS FROM THE CHILD TABLE
                let selected_documents = [];
                dialog.fields_dict.documents.grid.grid_rows.forEach(row => {
                    if (row && row.doc) {
                        selected_documents.push({
                            required_document: row.doc.required_document,
                            document_collection_stage: row.doc.document_collection_stage
                        });
                    }
                });

                // ? TRIGGER BACKEND METHOD WITH EXTRA INFO
                frappe.call({
                    method: "prompt_hr.py.utils.invite_for_document_collection",
                    args: {
                        args: {
                            name: frm.doc.job_applicant,
                            joining_document_checklist: values.joining_document_checklist,
                            document_collection_stage: values.document_collection_stage,
                            documents: selected_documents,
                        },
                        joining_document_checklist: values.joining_document_checklist,
                        document_collection_stage: values.document_collection_stage,
                        documents: selected_documents,
                        child_table_fieldname: "new_joinee_documents"
                    },
                    callback: function (r) {
                        if (r.message === "Already invited for document collection.") {
                            frappe.msgprint({
                                title: __("Info"),
                                message: __("This candidate has already been invited for document collection."),
                                indicator: "blue"
                            });
                        } else if (r.message === "An error occurred while inviting for document collection.") {
                            frappe.msgprint({
                                title: __("Error"),
                                message: __("Oops! Something went wrong while sending the invite. Please try again or check the logs."),
                                indicator: "red"
                            });
                        } else {
                            frappe.msgprint({
                                title: __("Success"),
                                message: __("Invite for document collection has been sent successfully!"),
                                indicator: "green"
                            });
                        }
                    }
                });
            }
        });

        dialog.show();
    });
}